Solutions/Ais.Net.Receiver/ReceiverHost.cs

```diff
@@ -0,0 +1,66 @@
+﻿// <copyright file="Program.cs" company="Endjin">
+// Copyright (c) Endjin. All rights reserved.
+// </copyright>
+
+namespace Ais.Net.Receiver
+{
+    using Microsoft.Extensions.Configuration;
+
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reactive.Linq;
+    using System.Threading.Tasks;
+
+    public class ReceiverHost
+    {
+        private readonly IConfiguration configuration;
+        private readonly IStorageClient storageClient;
+        private NmeaReceiver receiver;
+
+        public ReceiverHost(IConfiguration configuration, IStorageClient storageClient)
+        {
+            this.configuration = configuration;
+            this.storageClient = storageClient;
+        }
+
+        public async Task StartAsync()
+        {
+            this.receiver = new NmeaReceiver(
+                this.configuration[ConfigurationKeys.Host],
+                int.Parse(this.configuration[ConfigurationKeys.HostPort]),
+                TimeSpan.Parse(this.configuration[ConfigurationKeys.ConnectionRetryPeriodicity]),
+                int.Parse(this.configuration[ConfigurationKeys.ConnectionRetryAttempts]));
+
+            IObservable<string> messages = this.GetAsync().ToObservable();
+            await messages.Buffer(int.Parse(this.configuration[ConfigurationKeys.WriteBatchSize])).Select(this.storageClient.PersistAsync);
```

Comment:
Eew! If I'm reading this right, you're using the `Select` operator for its side effects?

Even if we ignore how wrong this is from mathematical elegance point of view, there's a serious practical problem here: `PersistAsync` returns a `Task` which you are ignoring. This causes two problems:

1. failures will go unnoticed, because nothing ever inspects the task returned by `PersistAsync`
2. you might end up with multiple concurrent batch writes in progress simultaneously

Even in a world without errors, 2 looks like it could be a serious problem here. You're using append operations on a blob, right? In blob storage, we group things by hour, right? I'm assuming `WriteBatchSize` here is expected to produce a chunk that is smaller than 1 hour, right? So in normal operation, shortly after the top of an hour you'll create a new blob, and then you expect to append to that blob multiple times over the course of the hour, right?

But nothing here will ensure that these appends happen in order. If you happen to receive enough of a splurge of data from the source (or if blob storage has a temporary perf issue) it's possible that you will start attempting a new append before the previous batch's append has finished.

I was going to say that you probably want to use Rx's `ForEachAsync` operator here, except it, ironically, doesn't actually have support for `async`... (Well, it sort of does. It returns something you can `await` so you've got an `async`-friendly way of saying "I want to wait until you've processed everything here" but the callback continues to be an `Action` and it doesn't have any `Func<Task>` forms to enable an async handler. The reason for that is that `IObserver.OnNext` is also non-async. Possibly one of Rx's async dimensions could work here—I think they have something like `IAsyncObservable<T>`, which I'm guessing would have this feature.)

However, maybe this is `IAsyncEnumerable<T>`'s time to shine. Instaed of calling `ToObservable` above, just bring in the library that adds the full set of Rx/Ix operators to `IAsyncEnumerable`, and this inclues the `Buffer` that you're using here. You could then write an `await foreach` over that, enabling you to enforce the "one batch at a time" behaviour that I'm pretty sure you want here.